File: Educative/SlidingWindow/LongestSubstringWithSameLettersAfterReplacement.py
# Given a string with lowercase letters only, if you are allowed to replace no more than k letters with any letter, find the length of the longest substring having the same letters after replacement.
# Time O(n), Space O(n)

def length_of_longest_substring(str1, k):
  window_start, max_window, max_freq = 0, 0, 0
  char_freq = {}

  for window_end in range(len(str1)):
    right_char = str1[window_end]
    char_freq[right_char] = char_freq.get(right_char, 0) + 1
    max_freq = max(max_freq, char_freq[right_char])

    if (window_end - window_start + 1) - max_freq > k:
      left_char = str1[window_start]
      char_freq[left_char] -= 1
      window_start += 1
    
    max_window = max(max_window, window_end - window_start + 1)
  return max_window

File: Educative/SlidingWindow/LongestSubstringWithSameLettersAfterReplacement.java
// Given a string with lowercase letters only, if you are allowed to replace no more than k letters with any letter, find the length of the longest substring having the same letters after replacement.
// Time O(n), Space O(n)

import java.util.*;

class CharacterReplacement {
    public static int findLength(String str, int k) {
      int windowStart = 0, maxLength = 0, maxRepeatCount = 0;
      Map<Character, Integer> freqMap = new HashMap<>();
  
      for (int windowEnd = 0; windowEnd < str.length(); windowEnd++){
        char rightChar = str.charAt(windowEnd);
        freqMap.put(rightChar, freqMap.getOrDefault(rightChar, 0) + 1);
        maxRepeatCount = Math.max(maxRepeatCount, freqMap.get(rightChar));
  
        if (windowEnd - windowStart + 1 - maxRepeatCount > k){
          char leftChar = str.charAt(windowStart);
          freqMap.put(leftChar, freqMap.get(leftChar) - 1);
          windowStart++;
        }
  
        maxLength = Math.max(maxLength, windowEnd - windowStart + 1);
      }
      return maxLength;
    }
  }